enhancements/update/upgrade-without-registry.md

```diff
@@ -0,0 +1,893 @@
+              "images": [
+                "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+                "quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...",
+                ...
+              ]
+            }
+```
+
+At that point the `bundle-extractor` job will finish.
+
+The `extracted` field for each node will indicate if the bundle extraction
+process has been completed.
+
+The `metadata` field for each node will contain the information from the
+`metadata.json` file of the bundle. Note that the list of images may be too long
+(approximately 180 images) to store it in the status of the `ClusterVersion`
+object; it may be convenient to store them in separate configmaps, and have only
+the references to those configmaps in the `ClusterVersion` status.
+
+When all the nodes have the bundle extracted (the `extracted` field for all
+nodes is `true`) the cluster version operator will delete the `bundle-server`
+daemon set and verify that the metadata in all nodes (the content of the
+`metadata` field) is the same. If there are differences they will be reported as
+error conditions in the status of the `ClusterVersion` object and the upgrade
+process will be aborted. This is intended to prevent accidents like having two
+different USB sticks with different bundles plugged in two different nodes.
+
+Once the metadata has been validated the cluster version operator will configure
+the machine config operator to pin the images specified in the metadata, so that
+they aren't removed by the garbage collection mechanisms while the upgrade is in
+progress. To do so it will use the new `pinnedImages` field of the
+`ContainerRuntimeConfig` object:
+
+```yaml
+apiVersion: machineconfiguration.openshift.io/v1
+kind: ContainerRuntimeConfig
+metadata:
+  name: pin-upgrade
+spec:
+  containerRuntimeConfig:
+    pinnedImages:
+    - quay.io/openshift-release-dev/ocp-release@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+    - quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:...
+```
+
+This configuration will be removed once the upgrade has completed successfully.
+
+`bundle-loader` job in each node of the cluster. The pods inside these jobs will
+first check that there is enough space available in the node to load the images
+into the `/var/lib/containers/storage` directory. To do so it will use the
+`size` field of the `metadata.json` file. Images in the bundle are compressed,
+but in `/var/lib/containers/storage` they are not, so the `bundle-loader` will
+check that there is at least twice the space indicated in the `size` field. If
+there is not enough space it will be reported as error conditions in the
+`CluterVersion` object and the upgrade process will be aborted.
+
+If there is enough space the `bundle-loader` pod will start a temporary embedded
+image registry server, listening in a randomly selected local port, using a self
```

Comment:
A host port?

Reply:
Yes, a host port, something like `localhost:12345` where `12345` is a randomly selected free port. This will be used only for the communication between CRI-O and the temporary registry, all within a single node.